proposals/0000-term-lookup-rules.md

```diff
@@ -0,0 +1,199 @@
+             |
+
+  namespace -> data
+             | type
+  ```
+   * With `data` specified in the import, only identifiers belonging to the data namespace will be brought into the scope.
+   * With `type` specified in the import, only identifiers belonging to the type namespace will be brought into the scope.
+* A new module `Data.Type` is introduced which defines unambiguous type definitions for built-in syntax constructors: `List`, `Tuple` and `Unit`:
+  ```haskell
+  type List = []
+  type Unit = ()
+  type Tuple2 = (,)
+  type Tuple3 = (,,)
+  type Tuple4 = (,,,)
+  {- ... -}
+
+  type family Tuple (xs :: [Type]) :: Type where
+    Tuple [] = Unit
+    Tuple [a] = GHC.Tuple.Unit a
+    Tuple [a,b] = Tuple2 a b
+    Tuple [a,b,c] = Tuple3 a b c
+    {- ... -}
+  ```
+* The `[a]` syntax means `[] a` when the `[]` type constructor is in scope, and a singleton list otherwise.
+* The `(a,b)` syntax means `(,) a b`, where `(,)` is either the type constructor or the data constructor according to the name lookup rules in the given context. This also applies to tuples of other arities.
+* `[]` and `()`, `(,)`, `(,,)`, ..., type constructors are no longer visible by default and instead are exported in a new module `Data.BuiltInTypes` which is imported by default like `Prelude`. `-XNoImplicitBuiltInTypes` is introduced to change this behavior.
+* Deprecate the `'` syntax of `-XDataKinds`, reserving this syntax for Template Haskell name quotation.
+* Deprecate the `''` syntax in Template Haskell.
+* Introduce a new warning, `-Wpun-declaration`, which warns the user when a declaration introduces a type constructor, type variable or a data constructor that clashes with an identifier in another namespace.
+* Introduce a new warning, `-Wpuns`, which warns the user of using an identifier that is found in both namespaces.
+## Examples
+
+Here are some examples of the new syntax:
+
+```haskell
+import Data.Proxy type as T
+import Data.Proxy data as D
+
+type T = T.Proxy Int
+
+f :: T
+f = D.Proxy
+```
+
+Warning examples:
+
+```haskell
+data T
+
+data A = T -- triggers -Wpun-declaration
+
+f = T -- triggers -Wpuns because T exists both in the type and the data namespaces
+```
+
+## Effect and Interactions
+
+* There is an asymmetry with what `-XDataKinds` does, as `-XDataKinds` only promotes data constructors to the type level and doesn't promote variables. On the contrary, new lookup rules let users reference type variables at the term-level. 
+  ```haskell
+  f :: forall a. a -> Int
+  f _ = sizeOf a -- valid code
+  ```
+  Compare this to `-XDataKinds`:
+  ```haskell
+  {-# LANGUAGE DataKinds #-}
+  a = 5
+  f :: Proxy a -> Proxy a -- 'a' here does not refer to the term-level 'a' and is implicitly quantified 
+  ```
+
+## Costs and Drawbacks
+
+* This proposal introduces new, potentially confusing for newcomers syntax, however, the newcomers can just look up the syntax.
+* It becomes impossible to refer to a punned name on type level:
```

Comment:
I don't understand this comment. Is this referring to the deprecation of `'`? Please say so.